src/bot.ts

```diff
@@ -552,36 +552,63 @@ async function withRetries<T>(
     task: () => Promise<T>,
     signal?: AbortSignal,
 ): Promise<T> {
+    // Set up delays between retries
+    const INITIAL_DELAY = 50; // ms
+    let lastDelay = INITIAL_DELAY;
+
+    // Define error handler
+    /**
+     * Determines the error handling strategy based on various error types.
+     * Sleeps if necessary, and returns whether to retry or rethrow an error.
+     */
+    async function handleError(error: unknown) {
+        let delay = false;
+        let strategy: "retry" | "rethrow" = "rethrow";
+
+        if (error instanceof HttpError) {
+            delay = true;
+            strategy = "retry";
+        } else if (error instanceof GrammyError) {
+            if (error.error_code >= 500) {
+                delay = true;
+                strategy = "retry";
+            } else if (error.error_code === 429) {
+                const retryAfter = error.parameters.retry_after;
+                if (retryAfter !== undefined) {
```

Comment:
not sure if this is better but this ensures that whatever the API returned is really a number. Its the strict inverse of checking what its not.
```suggestion
                if (typeof retryAfter === 'number') {
```